### PR TITLE
8251505: Use of types in compiler shared code should be consistent.

### PR DESCRIPTION
**Pull request by yftsai**
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8251505](https://bugs.openjdk.java.net/browse/JDK-8251505): Use of types in compiler shared code should be consistent.

### Reviewers
 * @eastig (no known github.com user name / role)
 * [Paul Hohensee](https://openjdk.java.net/census#phh) (@phohensee - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7294/head:pull/7294` \
`$ git checkout pull/7294`

Update a local copy of the PR: \
`$ git checkout pull/7294` \
`$ git pull https://git.openjdk.java.net/jdk pull/7294/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7294`

View PR using the GUI difftool: \
`$ git pr show -t 7294`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7294.diff">https://git.openjdk.java.net/jdk/pull/7294.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/utilities/elfFile.cpp [hunk 1]
```diff
@@ -91,7 +91,7 @@ int FileReader::read_buffer(void* buf, size_t size) {
   return fread(buf, 1, size, _fd);
 }
 
-bool FileReader::set_position(long offset) {
+bool FileReader::set_position(int64_t offset) {
```
Comment (eastig) [inline src/hotspot/share/utilities/elfFile.cpp:94]: You introduce a bug here.
`fseek` declaration:
```
int fseek ( FILE * stream, long int offset, int origin );
```

`fseek` will read only 32 bits of `offset` if `sizeof(long)==32`.

**Review by eastig — CHANGES_REQUESTED** [review]

**Review by phohensee — CHANGES_REQUESTED** [review]
Why not use 'jlong' everywhere instead of int64_t? You've got a mix of them (well, compileBroker.* uses jlong), better to be consistent.. Ditto INT64_FORMAT and JLONG_FORMAT. You'd also avoid declaring "declare_integer_type(int64_t)" in vmStructs.cpp.

In gc_globals.hpp, use of intx is likely intentional, so I'd leave it along. intx resolves to intptr_t (see globalDefinitions.hpp), which is 32 bits on 32-bit systems and 64 bits on 64-bit ones.

**Comment by dholmes-ora** [comment]
Do not use jlong everywhere. We should only be using jlong where we have values that will interact with Java code and so have to be jlong to be compatible. Otherwise for a 64-bit type we should be using int64_t, or uint64_t as appropriate. For a type that should be 32-bit or 64-bit depending on the environment we should use intptr_t or size_t depending on the nature of the variable.

Also do not force-push changes, just commit any changes as normal and push them. When the change is integrated the skara tooling will flatten things into one clean commit. If you force-push you mess up the PR review process.

Thanks.

**Comment by dean-long** [comment]
How about making the traversal mark type unsigned?

**Comment by phohensee** [comment]
Hi, David, I stand corrected. Is there a document somewhere about the policy, and has anyone gone through Hotspot to remove improper use of jlong?

So, belay my jlong suggestion, but now compileBroker.* should use int64_t. I think my gc_globals.hpp comment still stands.

**Comment by phohensee** [comment]
The traversal mark type is signed right now, so I'd leave it signed for this PR and file another one if we want to change it to unsigned. There are quite a few places where signed types are used for values that are intuitively unsigned. One reason I can think of to keep using signed types is that it's easy to detect overflow/wrap-around: just check the sign bit. Allows a bit of time to check for overflow/wrap-around without keeping an old value around.

**Review by phohensee — APPROVED** [review]

**Comment by eastig** [comment]
Other missed places to change:
```
jvmci/jvmciEnv.hpp:  long get_long_at(JVMCIPrimitiveArray array, int index);
services/memReporter.hpp:  inline long diff_in_current_scale(size_t s1, size_t s2) const {
services/memReporter.hpp:    long amount = (long)(s1 - s2);
services/memReporter.hpp:    long scale = (long)_scale;
services/memReporter.cpp:  long amount_diff = diff_in_current_scale(current_amount, early_amount);
services/memReporter.cpp:  long reserved_diff = diff_in_current_scale(current_reserved, early_reserved);
services/memReporter.cpp:  long committed_diff = diff_in_current_scale(current_committed, early_committed);
services/memReporter.cpp:      long overhead_diff = diff_in_current_scale(_current_baseline.malloc_tracking_overhead(),
services/memReporter.cpp:  long diff_used = diff_in_current_scale(current_stats.used(),
services/memReporter.cpp:  long diff_waste = diff_in_current_scale(current_waste, early_waste);
runtime/vmThread.cpp:  long interval_ms = SafepointTracing::time_since_last_safepoint_ms();
```

**Review by phohensee — APPROVED** [review]
Lgtm.

**Review by eastig — APPROVED** [review]
lgtm

**Comment by eastig** [comment]
> > Other missed places to change:
> > jvmci/jvmciEnv.hpp:  long get_long_at(JVMCIPrimitiveArray array, int index);
> > services/memReporter.hpp:  inline long diff_in_current_scale(size_t s1, size_t s2) const {
> > services/memReporter.hpp:    long amount = (long)(s1 - s2);
> > services/memReporter.hpp:    long scale = (long)_scale;
> > services/memReporter.cpp:  long amount_diff = diff_in_current_scale(current_amount, early_amount);
> > services/memReporter.cpp:  long reserved_diff = diff_in_current_scale(current_reserved, early_reserved);
> > services/memReporter.cpp:  long committed_diff = diff_in_current_scale(current_committed, early_committed);
> > services/memReporter.cpp:      long overhead_diff = diff_in_current_scale(_current_baseline.malloc_tracking_overhead(),
> > services/memReporter.cpp:  long diff_used = diff_in_current_scale(current_stats.used(),
> > services/memReporter.cpp:  long diff_waste = diff_in_current_scale(current_waste, early_waste);
> > runtime/vmThread.cpp:  long interval_ms = SafepointTracing::time_since_last_safepoint_ms();
> 
> Other than jvmci these are not "compiler shared code" - other cleanups in other areas will need their own RFE.
> 
> Cheers, David

Created:
https://bugs.openjdk.java.net/browse/JDK-8281213
https://bugs.openjdk.java.net/browse/JDK-8281214

**Review by dholmes-ora — COMMENTED** [review]
These changes also appear okay to me. Where we have changed from 32-bit to 64-bit types we will need to watch for issues with the 32-bit builds.

**Pull request closed without merging**